src/lightning/data/streaming/functions.py

```diff
@@ -0,0 +1,188 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from datetime import datetime
+from pathlib import Path
+from types import GeneratorType
+from typing import Any, Callable, Optional, Sequence, Union
+
+from lightning.data.streaming.constants import _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_46, _TORCH_GREATER_EQUAL_2_1_0
+from lightning.data.streaming.data_processor import DataChunkRecipe, DataProcessor, DataTransformRecipe, PrettyDirectory
+
+if _LIGHTNING_CLOUD_GREATER_EQUAL_0_5_46:
+    from lightning_cloud.resolver import _execute, _LightningSrcResolver
+
+if _TORCH_GREATER_EQUAL_2_1_0:
+    from torch.utils._pytree import tree_flatten
+
+
+def _get_input_dir(inputs: Sequence[Any]) -> str:
+    flattened_item, _ = tree_flatten(inputs[0])
+
+    indexed_paths = {
+        index: element
+        for index, element in enumerate(flattened_item)
+        if isinstance(element, str) and os.path.exists(element)
+    }
+
+    if len(indexed_paths) == 0:
+        raise ValueError(f"The provided item {inputs[0]} didn't contain any filepaths.")
+
+    absolute_path = str(Path(indexed_paths[0]).resolve())
+
+    if indexed_paths[0] != absolute_path:
+        raise ValueError("The provided path should be absolute.")
+
+    return "/" + os.path.join(*str(absolute_path).split("/")[:4])
+
+
+class LambdaDataTransformRecipe(DataTransformRecipe):
+    def __init__(self, fn: Callable[[str, Any], None], inputs: Sequence[Any]):
+        super().__init__()
+        self._fn = fn
+        self._inputs = inputs
+
+    def prepare_structure(self, input_dir: Optional[str]) -> Any:
+        return self._inputs
+
+    def prepare_item(self, output_dir: str, item_metadata: Any) -> None:  # type: ignore
+        self._fn(output_dir, item_metadata)
+
+
+class LambdaDataChunkRecipe(DataChunkRecipe):
+    def __init__(
+        self,
+        fn: Callable[[Any], None],
+        inputs: Sequence[Any],
+        chunk_size: Optional[int],
+        chunk_bytes: Optional[int],
+        compression: Optional[str],
+    ):
+        super().__init__(chunk_size=chunk_size, chunk_bytes=chunk_bytes, compression=compression)
+        self._fn = fn
+        self._inputs = inputs
+
+    def prepare_structure(self, input_dir: Optional[str]) -> Any:
+        return self._inputs
+
+    def prepare_item(self, item_metadata: Any) -> Any:  # type: ignore
+        if isinstance(self._fn, GeneratorType):
```

Comment:
we can probably do the same with the transform recipe above

Reply:
Withe the transform recipe you don't need to yield as there isn't any outputs supported just yet.